### PR TITLE
added support for controlling audio devices on linux

### PR DESCRIPTION
**Pull request by Pieterv24**
I've added support to control PA sinks directly with deej instead of only controlling individual applications

### CODE REVIEW

File: session_linux.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ type paSession struct {
 	sinkInputChannels byte
 }
 
-type masterSession struct {
+type paDevice struct {
```
Comment (omriharel) [inline session_linux.go:28]: Struct name is a bit inconsistent with other session structs. I suggest using `paDeviceSession` and correspondingly renaming `paSession` -> `paProcessSession`

File: session_linux.go [hunk 2]
```diff
@@ -60,15 +60,17 @@ func newPASession(
 	return s
 }
 
-func newMasterSession(
+func newPaDevice(
```
Comment (omriharel) [inline session_linux.go:63]: `newPADeviceSession` (abbreviations in name should be [all uppercase](https://github.com/golang/go/wiki/CodeReviewComments#initialisms)).

File: session_linux.go [hunk 3]
```diff
@@ -78,9 +80,17 @@ func newMasterSession(
 	var key string
 
 	if isOutput {
-		key = masterSessionName
```
Comment (omriharel) [inline session_linux.go:81]: Please simplify this logic:
```go
if isMaster {
    if isOutput {
        key = masterSessionName
    } else {
        key = inputSessionName
    }
} else {
    key = name
}

**Review by omriharel — COMMENTED** [review]
Hi @Pieterv24, looks good overall! I posted some minor comments.

In addition to the code changes, this functionality also needs to be documented properly in the README and config files (including removal of mentions of device-specific targeting being a Windows-only feature).

Please also provide users with an explanation for how they can find out the names for their audio devices (output and input), and what exactly they would need to specify to target them.

I'm also interested to know about the following:

- Can there be a conflict between device and process names? In Windows this never happens, because executables end with `.exe`
- Which OS and PA versions have you tested this with?

Thanks!

**Comment by Pieterv24** [comment]
I've updated the documentation.
As for whether it'd be possible for a conflict of device and process names to occur, technically I'd say yes although it is highly unlikely since device names are usually very unique, so someone would have to explicitly make a binary that has the same name as a output device.

There are a few courses of action we can take.
- Just leave it as is, this will most likely not cause any issues.
- Add something to check whether a session with the given key is already in the session list, if so, don't add a new one or add something like (1) to the end
- Use the Sink name instead of the Sink Description, this is more unique, but has the downside of being less human readable and less intuitive in use. for example `Built-in Audio Digital Stereo (IEC958)` would become `alsa_output.pci-0000_00_1f.3.iec958-stereo` which can only be retrieved by running `pactl list sinks` instead of being able to use the name that is used in the PulseAudio GUI.

I've tested this on Arch Linux, PulsaAudio 13.99.3 (Latest)

File: README.md [hunk 4]
```diff
@@ -41,8 +41,8 @@ deej is written in Go and [distributed](https://github.com/omriharel/deej/releas
 - Bind apps to different sliders
   - Bind multiple apps per slider (i.e. one slider for all your games)
   - Bind the master channel
+  - Bind specific audio devices by name (on Windows and Linux (_expirimental_))
```
Comment (omriharel) [inline README.md:44]: Small typo (`experimental`)

File: README.md [hunk 5]
```diff
@@ -104,8 +104,9 @@ noise_reduction: default
...
 - **_New:_** `deej.unmapped` is a special option to control all apps that aren't bound to any slider ("everything else") (_experimental_)
 - **_New:_** On Windows, `deej.current` is a special option to control whichever app is currently in focus (_experimental_)
-- On Windows, you can specify a device's full name, i.e. `Speakers (Realtek High Definition Audio)`, to bind that device's level to a slider. This doesn't conflict with the default `master` and `mic` options, and works for both input and output devices.
-  - Be sure to use the full device name, as seen in the menu that comes up when left-clicking the speaker icon in the tray menu
+- You can specify a device's full name, i.e. `Speakers (Realtek High Definition Audio)`, to bind that device's level to a slider. This doesn't conflict with the default `master` and `mic` options, and works for both input and output devices.
+  - Windows: Use the full device name, as seen in the menu that comes up when left-clicking the speaker icon in the tray menu
+  - Linux (_expirimental_): Use the full device name, as seen in the `Output Devices` tab in `pavucontrol` or the description when running `pactl list sinks`
```
Comment (omriharel) [inline README.md:109]: - Typo (same as above)
- I would prefer to distinguish the difference by writing something like:
```
Use the device's description, as seen in PulseAudio's output devices tab (or by running `pactl list sinks`)
```

File: config.yaml [hunk 6]
```diff
@@ -1,9 +1,9 @@
 # process names are case-insensitive
 # you can use 'master' to indicate the master channel, or a list of process names to create a group
 # you can use 'mic' to control your mic input level (uses the default recording device)
+# you can use a device's full name, i.e. "Speakers (Realtek High Definition Audio)", to bind it. this works for both output and input devices
```
Comment (omriharel) [inline config.yaml:4]: - Please distinguish between windows and linux here as well. It doesn't need to be as thorough as the explanation in `README.md`, but it should mention that for linux it uses the device's description
- The changes to this file should be replicated to the config template file under `scripts`

File: session_linux.go [hunk 7]
```diff
@@ -79,18 +79,16 @@ func newPaDevice(
 
 	var key string
 
-	if isOutput {
-		if isMaster {
+	// If device is the master, check if its the mic or speaker and give apporpriate default name
```
Comment (omriharel) [inline session_linux.go:82]: Typo: `appropriate`

Comment (omriharel) [reply session_linux.go:82]: (although this is a fairly obvious `if` statement, I'm not sure it needs commenting in the first place..)

File: session_linux.go [hunk 8]
```diff
@@ -25,7 +25,7 @@ type paSession struct {
 	sinkInputChannels byte
 }
 
-type paDevice struct {
+type paDeviceSession struct {
```
Comment (omriharel) [inline session_linux.go:28]: Great, but what about `paSession`? Needs to be changed too for consistency

**Review by omriharel — CHANGES_REQUESTED** [review]
@Pieterv24, please see comments. Thanks!

**Comment by fkreiner** [comment]
Is this still a Thing?

**Comment by omriharel** [comment]
@fkreiner it is, but I'm going to need to find time to bring it up to speed w/ my requirements. I keep putting it off, but I'll prioritize it higher now that there's more interest.

**Comment by fkreiner** [comment]
No pressure. Just wanted to see if pulse will be implemented some day

**Comment by omriharel** [comment]
Just to clarify, pulse is already supported and you can build and use deej
on Linux right now (check out the main readme for instructions).

This PR further adds to that with device targeting capabilities, which are
currently missing on Linux (being able to tell deej to control a specific
audio device like speakers or headphones, as opposed to just apps/master).

**Comment by oerkel47** [comment]
This would be a nice addition! :-)

**Pull request closed without merging**